SYCL/ESIMD/api/functional/logger.hpp

```diff
@@ -0,0 +1,53 @@
+//===-- logger.hpp - Logger struct declaration ----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the declaration of the Logger struct, which using for
+/// printing verbose information ot the console.
+///
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
+#include <string>
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+namespace logger {
+
+// Class that provide methods for printing verbose messages to the console
+struct Logger {
+  Logger() = default;
+
+  // Notify a test has failed. Printed to the console provided message also
+  // print data type that was using when test was failed and number of
+  // elements that was using when test was failed if they was provided to the
+  // this function
+  template <int NumElems = 0>
+  static void fail(const std::string &msg,
+                   std::string underlying_data_type = "") {
+    std::string log_msg{msg};
+    if (underlying_data_type != "") {
```

Comment:
```suggestion
    if (!underlying_data_type.empty()) {
```
I have a feeling such API will be a bottle-neck for the following tests, so we might need to provide more generic interface someday..

Reply:
this comment was applied in this commit: 98d99192e62693a3205e15f51dc5a4bd3656f338